### PR TITLE
Fix Build/Update Boost on Windows (non-PR-754 version)

### PR DESCRIPTION
**Pull request by stephengtuggy**
Thank you for submitting a pull request and becoming a contributor to the Vega Strike Core Engine.

Please answer the following:

Code Changes:
- [X] Have the PR Validation Tests been run? See https://github.com/vegastrike/Vega-Strike-Engine-Source/wiki/Pull-Request-Validation

Issues:
- #782 
- #783 

Purpose:
- What is this pull request trying to do? Update Boost on Windows, on the `master` branch.
- What release is this for? 0.9.x
- Is there a project or milestone we should apply this to? 0.9.x

### CODE REVIEW

File: engine/src/cmd/collide_map.cpp
Comment (stephengtuggy) [inline engine/src/cmd/collide_map.cpp:58]: Please note that this change may affect game behavior, unlike most of the changes in this PR. However, I think this block of changes is almost mandatory for correctness.

**Comment by stephengtuggy** [comment]
Please note also that this looks like a very large changed-line count, but probably at least 1,000 of those are just whitespace changes.

**Comment by BenjamenMeyer** [comment]
@stephengtuggy Try https://github.com/boostorg/python/issues/233 and see if that will do it

**Comment by stephengtuggy** [comment]
This was working, and everything was golden. Then nasm.us went down (https://github.com/microsoft/vcpkg/issues/32600), and all our Windows builds stopped working. Dang it!

**Comment by stephengtuggy** [comment]
I think I'm going to go ahead and submit this PR for review anyway, based on the fact that all the builds were succeeding prior to nasm.us going down and breaking the openssl builds.

**Comment by stephengtuggy** [comment]
@evertvorster , if you have time to play-test this PR a bit before it gets merged, that would be great. If not, no worries.

File: engine/src/cmd/collide_map.h
Comment (BenjamenMeyer) [inline engine/src/cmd/collide_map.h:203]: Yes, even if just empty and/or calling the parent

File: script/bootstrap.ps1
Comment (BenjamenMeyer) [inline script/bootstrap.ps1:34]: I assume this is temporary while the service is down, no?

**Review by BenjamenMeyer — COMMENTED** [review]
Couple questions before I approve

File: engine/src/cmd/collide_map.h
Comment (stephengtuggy) [reply engine/src/cmd/collide_map.h:203]: Agreed. That is most likely what needs to happen, sooner or later. But can it wait for another PR?

File: script/bootstrap.ps1
Comment (stephengtuggy) [reply script/bootstrap.ps1:34]: > I assume this is temporary while the service is down, no?

@BenjamenMeyer Actually, no. This is intended to be a long-term solution. For the last several months -- well before the nasm.us issues -- I have found it necessary to use a fork-of-a-fork of vcpkg in order to get Vega Strike to build successfully on Windows. I do intend to keep this fork updated semi-regularly.

Comment (stephengtuggy) [reply script/bootstrap.ps1:34]: @BenjamenMeyer working on that right now: https://github.com/vegastrike/vcpkg-local

File: engine/src/cmd/collide_map.h
Comment (BenjamenMeyer) [reply engine/src/cmd/collide_map.h:203]: yes, we probably need to do a PR with adding it to everything we can find; though adding:
```c++
virtual ~CollideMap() { ~ListMutableSet<Collidable>() }
```
here would do well too.

File: engine/src/cmd/collide_map.cpp
Comment (BenjamenMeyer) [inline engine/src/cmd/collide_map.cpp:165]: we should just test directly against `.end()` instead of creating a variable here.

**Review by BenjamenMeyer — CHANGES_REQUESTED** [review]

File: engine/src/cmd/collide_map.cpp
Comment (BenjamenMeyer) [reply engine/src/cmd/collide_map.cpp:165]: NOTE: the `end` could potentially change which is why you're always supposed to test directly against `end()` instead of using a var

File: .github/workflows/Windows-CI.yml
Comment (royfalk) [inline .github/workflows/Windows-CI.yml:22]: Are we really testing all these variants? It may be a bit excessive IMO.

File: .github/workflows/gh-actions-pr.yml
Comment (royfalk) [inline .github/workflows/gh-actions-pr.yml:39]: Again, I think we need to restrict our officially supported everything. We are not a large corporation and even there I don't recall supporting so many versions.

File: .github/workflows/gh-actions-pr.yml
Comment (royfalk) [inline .github/workflows/gh-actions-pr.yml:304]: Also, this looks way more than just updating boost. It's better and easier to submit several smaller PRs than a big one.

File: engine/CMakeLists.txt
Comment (royfalk) [inline engine/CMakeLists.txt:929]: Easier to have these cosmetic (tabs and copyright changes) in a separate PR. It's pretty much auto approve and removes a lot of noise from here.
Also, probably a good idea to squash the commits to keep the tree cleaner.

File: engine/src/cmd/base_util.cpp
Comment (royfalk) [inline engine/src/cmd/base_util.cpp:31]: Same as above.

File: engine/src/cmd/collide_map.cpp
Comment (royfalk) [inline engine/src/cmd/collide_map.cpp:54]: I don't understand this code at all. Can we maybe split this into a separate PR?

File: engine/src/cmd/collide_map.cpp
Comment (royfalk) [inline engine/src/cmd/collide_map.cpp:163]: I'm fine with type inference where it is trivial. Example from other languages is something like

`let someString = String('blah blah')`

However, I'm less comfortable with replacing _std::list<CollidableBackref>::iterator_ with _auto_. I prefer to clearly see my C++ abominations.

File: engine/src/cmd/collide_map.h
Comment (royfalk) [inline engine/src/cmd/collide_map.h:46]: What does this do? I assume it's defined either way.
Also, can we fix the indentation. It's handing out there.

File: engine/src/python/unit_exports.cpp
Comment (royfalk) [inline engine/src/python/unit_exports.cpp:24]: Here as well.

File: engine/src/python/unit_wrapper_class.h
Comment (royfalk) [inline engine/src/python/unit_wrapper_class.h:165]: Probably better to just not break the line or break earlier for easier reading. Same for the rest of the file.

File: engine/src/universe_util_generic.cpp
Comment (royfalk) [inline engine/src/universe_util_generic.cpp:731]: Did you move this code block to an inner brace intentionally?

File: engine/src/cmd/collide_map.h
Comment (stephengtuggy) [reply engine/src/cmd/collide_map.h:203]: I don't think virtual destructors need to call each other. IIRC, they automatically get called in the proper sequence.

Making the destructor virtual, for every class that is intended to be inheritable, IS something that needs to happen. But I think in this case, I'll wait until the next PR.

Thanks for the review!

File: script/bootstrap.ps1
Comment (stephengtuggy) [reply script/bootstrap.ps1:34]: Continuing to have problems with the Windows builds. Now it's Python that isn't compiling, via vcpkg. (Annoying that we even have to compile it at all. Maybe there's another way; not sure.)

File: engine/src/cmd/collide_map.cpp
Comment (stephengtuggy) [reply engine/src/cmd/collide_map.cpp:165]: @BenjamenMeyer you're probably right. I would like to point out that the `.end()` pre-assignment was there before; I just changed the variable type declarations to `auto`, based on CLion's suggestion to use `auto` for all iterator variables. Based on @royfalk 's comment further down, it sounds like we still need to discuss whether or not we want to do that as a general practice. Maybe I'll revert the changes to these lines for now.

File: .github/workflows/gh-actions-pr.yml
Comment (stephengtuggy) [reply .github/workflows/gh-actions-pr.yml:304]: Normally, I would agree. In this case, I was just trying to hurry up and get everything building again.

Good point that this is about a lot more than updating Boost. I have updated the PR title to reflect that.

File: .github/workflows/Windows-CI.yml
Comment (stephengtuggy) [reply .github/workflows/Windows-CI.yml:22]: Good point. We can probably cut back a bit.

File: engine/CMakeLists.txt
Comment (stephengtuggy) [reply engine/CMakeLists.txt:929]: I can squash the commits when merging this PR. I am fine with that.

Do you want me to reverse these cosmetic changes to CMakeLists.txt in this PR, and then reformat later in a separate PR?

File: engine/src/cmd/base_init.cpp
Comment (stephengtuggy) [reply engine/src/cmd/base_init.cpp:31]: @royfalk Indeed, this is quite an ancient version of Boost. However... there is a possibility that this specific version is being used to distinguish between Python environments (embedded Python vs. ... doubly-embedded Python, or something like that). I'm not quite sure of the details or history of this. Would like to get some input from a long-term project member or two, if possible.

File: .github/workflows/gh-actions-pr.yml
Comment (stephengtuggy) [reply .github/workflows/gh-actions-pr.yml:39]: I agree. The policy we decided on last was to support the LTS versions of distros such as Ubuntu that are still in mainstream support, as well as the latest. We may have to revisit that a bit; it's turning out to be more than I can handle lately.

File: engine/src/cmd/base_init.cpp
Comment (stephengtuggy) [reply engine/src/cmd/base_init.cpp:31]: In the meantime, can we leave it as is?

File: engine/src/cmd/collide_map.cpp
Comment (stephengtuggy) [reply engine/src/cmd/collide_map.cpp:54]: We were doing a `static_cast` to pointer-to-CollidableBackref, without any type checking, where we should have been doing a `dynamic_cast` in order to verify that the actual type (at runtime) is correct.

Does that explain it sufficiently? Or do you want to go ahead and split this part out into its own PR?

File: engine/src/cmd/collide_map.cpp
Comment (stephengtuggy) [reply engine/src/cmd/collide_map.cpp:163]: Fair point LOL.

I was going by a CLion suggestion that all iterator variables should use `auto`. Sounds like we need to discuss that further though.

File: engine/src/cmd/collide_map.h
Comment (stephengtuggy) [reply engine/src/cmd/collide_map.h:46]: Adding the empty braces initializes the value of `ref`. It was uninitialized previously, as far as I can tell.

File: engine/src/universe_util_generic.cpp
Comment (stephengtuggy) [reply engine/src/universe_util_generic.cpp:731]: Let me take another look. I think so, but now I'm not sure which way is correct.

File: engine/src/universe_util_generic.cpp
Comment (stephengtuggy) [reply engine/src/universe_util_generic.cpp:685]: Yeah, I'm using CLion. I highly reco

File: engine/src/cmd/collide_map.cpp
Comment (royfalk) [reply engine/src/cmd/collide_map.cpp:54]: That's fine.

File: engine/CMakeLists.txt
Comment (royfalk) [reply engine/CMakeLists.txt:929]: No. It's already in review and I'm not that strict about full separation. Everything in moderation.

File: engine/src/cmd/base_init.cpp
Comment (royfalk) [reply engine/src/cmd/base_init.cpp:31]: Sure.

File: .github/workflows/gh-actions-pr.yml
Comment (BenjamenMeyer) [reply .github/workflows/gh-actions-pr.yml:39]: It's generally best effort right now. If we need to drop something because we don't have the bandwidth then that's fine - let's just document it and why.

That said, most of our distros are highly related with some simple packaging changes between them so I don't think it's a huge effort for Linux. Mac/Windows, OTOH, are still in a larger need of support and we could certainly use some help there.

File: engine/src/cmd/collide_map.h
Comment (BenjamenMeyer) [reply engine/src/cmd/collide_map.h:46]: C++11 initialization method; we've generally stuck to constructor initialization but we've started getting more C++11 brace initialization in. We just have to make sure it's C++11 format and not C++20 feature in the process.

File: engine/src/cmd/collide_map.cpp
Comment (BenjamenMeyer) [reply engine/src/cmd/collide_map.cpp:58]: yeah, the code on the left certainly isn't right. I did some similar change in @royfalk 's SDL2 PR....https://github.com/royfalk/Vega-Strike-Engine-Source/pull/1/files#r1292449185

File: engine/src/cmd/collide_map.cpp
Comment (BenjamenMeyer) [reply engine/src/cmd/collide_map.cpp:165]: yeah,  I realize it was there before; this would be one improvement for the code changing that up to actually be correct. Use `auto` instead of specifying the full iterator type is also useful here (and one of few things I like to use `auto` for).

File: engine/src/cmd/collide_map.h
Comment (stephengtuggy) [reply engine/src/cmd/collide_map.h:203]: Filed an issue for this. Resolving conversation for now.

File: engine/src/cmd/collide_map.cpp
Comment (stephengtuggy) [reply engine/src/cmd/collide_map.cpp:165]: OK, I'll change this to use `.end()` directly. I'll also leave it as `auto` for now. Any objections?

File: .github/workflows/gh-actions-pr.yml
Comment (stephengtuggy) [reply .github/workflows/gh-actions-pr.yml:39]: Definitely.

File: engine/src/cmd/base_init.cpp
Comment (stephengtuggy) [reply engine/src/cmd/base_init.cpp:31]: Filed as #797

File: engine/src/cmd/base_util.cpp
Comment (stephengtuggy) [reply engine/src/cmd/base_util.cpp:31]: Covered by #797

File: engine/src/cmd/collide_map.cpp
Comment (stephengtuggy) [reply engine/src/cmd/collide_map.cpp:163]: Started a discussion for this: https://github.com/orgs/vegastrike/discussions/800

File: engine/src/cmd/collide_map.h
Comment (royfalk) [reply engine/src/cmd/collide_map.h:46]: Intuitively, it seemed to me it was initialized. However, on closer inspection, this is a non-static variable, so it really is uninitialized. it would be clearer if we were to specify :

`Unit *unit = nullptr;
        unsigned int bolt_index;
 } ref;`

Comment (BenjamenMeyer) [reply engine/src/cmd/collide_map.h:46]: This initialization is for any kind of variable, not simply non-static. The C++ committee's idea is that the constructors don't need to do all the initialization of the class since they have this other means.
I think Strostroup's idea is that the constructor isn't necessary in 99% of cases and this lets them do that.

That said, I still kind of like having explicit constructors doing all the initialization.

File: engine/src/cmd/collide_map.cpp
Comment (BenjamenMeyer) [reply engine/src/cmd/collide_map.cpp:163]: I'm okay with `auto` for the loops with iterators because (a) the loops should be fairly short, (b) the iterator notation (`namespace::type::iterator`) can get quite long and lengthy and doesn't serve much purpose, and (c) the iterator type can change somewhat so it's helpful.

So this is actually one of the few places I think `auto` is okay. That said, I already mentioned that the `end` check needs to be done in the loop since `end` could  change or become invalid during the loop.

File: engine/src/cmd/collide_map.cpp
Comment (BenjamenMeyer) [reply engine/src/cmd/collide_map.cpp:165]: I think this is all that this PR is waiting on for approval

**Comment by BenjamenMeyer** [comment]
@stephengtuggy please merge in master

File: engine/src/cmd/collide_map.cpp
Comment (stephengtuggy) [reply engine/src/cmd/collide_map.cpp]: Fixed.

**Comment by stephengtuggy** [comment]
> @stephengtuggy please merge in master

Done, I believe.

File: .github/workflows/Windows-CI.yml
Comment (BenjamenMeyer) [reply .github/workflows/Windows-CI.yml:22]: let's just stick to 1 Windows build for the time being; that should at least get the ball rolling.

Comment (stephengtuggy) [reply .github/workflows/Windows-CI.yml:22]: How about 2 Windows builds? One on Windows (Server) 2019, and one on Windows (Server) 2022. I think that will be enough even to satisfy me, haha.

**Comment by BenjamenMeyer** [comment]
> > @stephengtuggy please merge in master
> 
> Done, I believe.

Unfortunately not, or at least the merge didn't work right. It's failing at minimum because it can't find SDL2; and checking the bootstrap script it doesn't seem to be trying to install it so it doesn't have some of the required changes.
It's almost acting like it has the wrong Docker images too. Hopefully it's just the bootstrap scripts and some related CMake settings for SDL2.

File: .github/workflows/Windows-CI.yml
Comment (BenjamenMeyer) [reply .github/workflows/Windows-CI.yml:22]: @stephengtuggy if we need to

**Comment by stephengtuggy** [comment]
> > > @stephengtuggy please merge in master
> > 
> > 
> > Done, I believe.
> 
> Unfortunately not, or at least the merge didn't work right. It's failing at minimum because it can't find SDL2; and checking the bootstrap script it doesn't seem to be trying to install it so it doesn't have some of the required changes. It's almost acting like it has the wrong Docker images too. Hopefully it's just the bootstrap scripts and some related CMake settings for SDL2.

@BenjamenMeyer I merged your PR into mine. 😃 Hopefully it works now. Thanks.

**Comment by BenjamenMeyer** [comment]
Well more successes than failures overall.
- Windows 2022 passed; 2019 did not
- most Mac builds failed, but 3 passed
- all Linux builds passed except Ubuntu Kinetic
So progress

**Comment by stephengtuggy** [comment]
(Yes, I know that's a change in tune from what I usually say 😄 )

**Comment by BenjamenMeyer** [comment]
@stephengtuggy I can live with that.

File: .github/workflows/Windows-CI.yml
Comment (BenjamenMeyer) [reply .github/workflows/Windows-CI.yml:22]: well, VS2022 on Windows 2022 passed.

File: .github/workflows/gh-actions-pr.yml
Comment (BenjamenMeyer) [reply .github/workflows/gh-actions-pr.yml:39]: So Kinetic is no longer supported since Manic is out. Will fix in master.

File: .github/workflows/macos-ci.yml
Comment (BenjamenMeyer) [inline .github/workflows/macos-ci.yml:53]: NOTE: All the MacOS 11 builds failed; MacOS 12 passed for all except including OpenAL with GCC if I understand the matrix right.

**Review by BenjamenMeyer — APPROVED** [review]
We'll fix the Mac/Windows builds and swap Ubuntu Kinetic for Mantic on master.

**Comment by stephengtuggy** [comment]
Merged, finally!

File: engine/src/universe_util_generic.cpp
Comment (stephengtuggy) [reply engine/src/universe_util_generic.cpp:731]: Update: Yes, this was intentional. I think a previous PR of mine messed up the structure of this file a bit. This is an attempt to fix it.

File: .github/workflows/macos-ci.yml
Comment (stephengtuggy) [reply .github/workflows/macos-ci.yml:53]: Filed https://github.com/vegastrike/Vega-Strike-Engine-Source/issues/812 for the macOS build failures. They all seem to be related to which version of Python Boost is trying to use.

**Pull request merged**
